File: app/src/main/java/me/zhanghai/android/files/settings/DefaultIfEmptyEditTextPreference.java
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.settings;

import android.content.Context;
import android.content.res.TypedArray;
import android.text.TextUtils;
import android.util.AttributeSet;

import com.takisoft.preferencex.EditTextPreference;

import androidx.annotation.AttrRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StyleRes;

public class DefaultIfEmptyEditTextPreference extends EditTextPreference {

    private String mDefaultValue;

    public DefaultIfEmptyEditTextPreference(@NonNull Context context) {
        super(context);
    }

    public DefaultIfEmptyEditTextPreference(@NonNull Context context,
                                            @Nullable AttributeSet attrs) {
        super(context, attrs);
    }

    public DefaultIfEmptyEditTextPreference(@NonNull Context context, @Nullable AttributeSet attrs,
                                            @AttrRes int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    public DefaultIfEmptyEditTextPreference(@NonNull Context context, @Nullable AttributeSet attrs,
                                            @AttrRes int defStyleAttr, @StyleRes int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
    }

    @Override
    protected Object onGetDefaultValue(@NonNull TypedArray a, int index) {
        mDefaultValue = (String) super.onGetDefaultValue(a, index);
        return mDefaultValue;
    }

    @Override
    public void setDefaultValue(@Nullable Object defaultValue) {
        super.setDefaultValue(defaultValue);
        mDefaultValue = (String) defaultValue;
    }

    @Override
    public void setText(@Nullable String text) {
        if (TextUtils.isEmpty(text)) {
            text = mDefaultValue;
        }
        super.setText(text);
    }
}

File: app/src/main/java/me/zhanghai/android/files/settings/DefaultIfEmptyEditTextPreference.kt
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.settings

import android.content.Context
import android.content.res.TypedArray
import android.util.AttributeSet
import androidx.annotation.AttrRes
import androidx.annotation.StyleRes
import com.takisoft.preferencex.EditTextPreference

class DefaultIfEmptyEditTextPreference : EditTextPreference {
    private var defaultValue: String? = null

    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs)

    constructor(context: Context, attrs: AttributeSet?, @AttrRes defStyleAttr: Int) : super(
        context, attrs, defStyleAttr
    )

    constructor(
        context: Context,
        attrs: AttributeSet?,
        @AttrRes defStyleAttr: Int,
        @StyleRes defStyleRes: Int
    ) : super(context, attrs, defStyleAttr, defStyleRes)

    override fun onGetDefaultValue(a: TypedArray, index: Int): Any? =
        super.onGetDefaultValue(a, index).also { defaultValue = it as String? }

    override fun setDefaultValue(defaultValue: Any?) {
        super.setDefaultValue(defaultValue)

        this.defaultValue = defaultValue as String?
    }

    override fun setText(text: String?) {
        val text = if (!text.isNullOrEmpty()) text else defaultValue
        super.setText(text)
    }
}
